### PR TITLE
[FLINK-9187][METRICS] add prometheus pushgateway reporter

### PR DESCRIPTION
**Pull request by lamber-ken**
## What is the purpose of the change
This pull request makes flink system can send metrics to prometheus via pushgateway. it may be useful.

## Brief change log

  - Add prometheus pushgateway repoter
  - Restructure the code of the promethues reporter part

## Verifying this change

This change is already covered by existing tests. [prometheus test](https://github.com/apache/flink/tree/master/flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by lamber-ken** [comment]
Please click here for details [old-flink-9187](https://github.com/apache/flink/pull/5857)

File: docs/monitoring/metrics.md [hunk 1]
```diff
@@ -700,6 +700,31 @@ Flink metric types are mapped to Prometheus metric types as follows:
...
+
+metrics.reporter.promgateway.class: org.apache.flink.metrics.prometheus.PrometheusPushGatewayReporter
+metrics.reporter.promgateway.host: localhost
+metrics.reporter.promgateway.port: 9091
+metrics.reporter.promgateway.prefix: flink
+
+{% endhighlight %}
+
+PrometheusPushGatewayReporter push metrics to a [Pushgateway](https://github.com/prometheus/pushgateway). The Pushgateway then exposes 
+these metrics to Prometheus. The working mechanism is different from PrometheusReporter (see [PrometheusReporter](#prometheus-orgapacheflinkmetricsprometheusprometheusreporter)).
```
Comment (zentol) [inline docs/monitoring/metrics.md]: I will remove the second sentence as it provides no value, and instead introduce a link tot he prometheus documentation for pushgateway use-cases.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		} catch (Exception e) {
+			log.warn("Failed reporting metrics to Prometheus.", e);
+		}
+	}
+
+	@Override
+	public void close() {
+		if (pushGateway != null) {
+			try {
+				pushGateway.delete(jobName);
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: I will add a configuration option to make the deletion optional, users may want to be in control of this themselves.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		String jobNamePrefix = config.getString(ARG_JOB_NAME_PREFIX, DEFAULT_JOB_NAME_PREFIX);
+
+		// host port
+		if (host == null || host.length() == 0 || port < 1) {
+			throw new IllegalArgumentException("Invalid host/port configuration. Host: " + host + " Port: " + port);
+		}
+
+		// jobname
+		String random = new AbstractID().toString();
+		jobName = jobNamePrefix + JOB_NAME_SEPARATOR + random;
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: I will make the random suffix optional for cases where users can supply different configurations to each container.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	@Override
+	public void open(MetricConfig config) {
+
+		// reporter configs
+		String host = config.getString(ARG_HOST, null);
+		int port = config.getInteger(ARG_PORT, -1);
+		String jobNamePrefix = config.getString(ARG_JOB_NAME_PREFIX, DEFAULT_JOB_NAME_PREFIX);
+
+		// host port
+		if (host == null || host.length() == 0 || port < 1) {
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: use `host.isEmpty()`

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.exporter.PushGateway;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * /**
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: remove line

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/AbstractPrometheusReporter.java [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+
+/**
+ * base prometheus reporter for prometheus metrics.
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/AbstractPrometheusReporter.java]: extend javadocs to include responsibilities of this/extending classes.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusReporter.java [hunk 7]
```diff
@@ -20,76 +20,38 @@
...
-import java.util.regex.Pattern;
 
 /**
  * {@link MetricReporter} that exports {@link Metric Metrics} via Prometheus.
  */
 @PublicEvolving
-public class PrometheusReporter implements MetricReporter {
-	private static final Logger LOG = LoggerFactory.getLogger(PrometheusReporter.class);
+public class PrometheusReporter extends AbstractPrometheusReporter {
+	private final Logger log = LoggerFactory.getLogger(PrometheusReporter.class);
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusReporter.java]: remove this logger and use the logger defined in the parent class

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+		String host = config.getString(ARG_HOST, null);
+		int port = config.getInteger(ARG_PORT, -1);
+		String jobNamePrefix = config.getString(ARG_JOB_NAME_PREFIX, DEFAULT_JOB_NAME_PREFIX);
+
+		// host port
+		if (host == null || host.length() == 0 || port < 1) {
+			throw new IllegalArgumentException("Invalid host/port configuration. Host: " + host + " Port: " + port);
+		}
+
+		// jobname
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: remove comment, adds no value

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+	@Override
+	public void open(MetricConfig config) {
+
+		// reporter configs
+		String host = config.getString(ARG_HOST, null);
+		int port = config.getInteger(ARG_PORT, -1);
+		String jobNamePrefix = config.getString(ARG_JOB_NAME_PREFIX, DEFAULT_JOB_NAME_PREFIX);
+
+		// host port
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: remove comment, adds no value

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	public static final char JOB_NAME_SEPARATOR = '-';
+	public static final String DEFAULT_JOB_NAME_PREFIX = "flink";
+
+	private PushGateway pushGateway;
+	private String jobName;
+
+	@Override
+	public void open(MetricConfig config) {
+
+		// reporter configs
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: remove comment, adds no value

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+
+	@Override
+	public void close() {
+		if (pushGateway != null) {
+			try {
+				pushGateway.delete(jobName);
+			} catch (IOException e) {
+				log.warn("Failed to delete the job of Pushgateway", e);
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: include job name.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+		pushGateway = new PushGateway(host + ":" + port);
+		log.info("Configured PrometheusPushGatewayReporter with {host:{}, port:{}, jobName:{}}", host, port, jobName);
+	}
+
+	@Override
+	public void report() {
+		try {
+			pushGateway.push(CollectorRegistry.defaultRegistry, jobName);
+		} catch (Exception e) {
+			log.warn("Failed reporting metrics to Prometheus.", e);
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: include job name

Comment (zentol) [reply flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java]: sync the exception message with exception in `close`

**Comment by lamber-ken** [comment]
@zentol,  well done. If there is anything I can do for you, please let me know.

**Pull request closed without merging**